### PR TITLE
only allow 1 user account for a single github account 

### PR DESCRIPTION
**Pull request by house9**
see issue see #25

### CODE REVIEW

File: lib/ppwm_matcher/models/user.rb [hunk 1]
```diff
@@ -24,14 +24,19 @@ def self.current(github_user)
       where(:github_login => github_user.login).limit(1).first
     end
 
+    # now this method is really doing find_or_create_or_update_email - doh!
     def self.find_or_create(email, github_user)
-      user = where(:email => email).limit(1).first
-      unless user
+      user = where(:github_login => github_user.login).limit(1).first
```
Comment (rwjblue) [inline lib/ppwm_matcher/models/user.rb]: Is it possible to use first_or_create to refactor this `if code` conditional.  Something like:

``` ruby
    def self.find_or_create(email, github_user)
      user = where(:github_login => github_user.login).first_or_create(
        :gravatar_id   => github_user.gravatar_id,
        :github_login  => github_user.login
      )

      user.update_attributes(:email, email) # calls save directly

      user
    end
```

File: lib/ppwm_matcher/models/user.rb [hunk 2]
```diff
@@ -24,14 +24,19 @@ def self.current(github_user)
       where(:github_login => github_user.login).limit(1).first
     end
 
+    # now this method is really doing find_or_create_or_update_email - doh!
```
Comment (bf4) [inline lib/ppwm_matcher/models/user.rb]: update or create?

File: lib/ppwm_matcher/models/user.rb (hunk 1 above)
Comment (bf4) [inline lib/ppwm_matcher/models/user.rb]: wouldn't that save it twice?  isn't there a :first_or_initialize? at least then you can user.email = email && user.save to hit the db one less time

Comment (house9) [inline lib/ppwm_matcher/models/user.rb]: just tried it - breaks a few tests
i could give it another try after lunch

Comment (rwjblue) [inline lib/ppwm_matcher/models/user.rb]: `update_attributes` only saves if there is a change.

Comment (rwjblue) [inline lib/ppwm_matcher/models/user.rb]: `user.update_attributes(:email, email)` should be `user.update_attributes(:email => email)`

I forgot that `update_attribute` takes two params where `update_attributes` takes a hash.

Comment (bf4) [inline lib/ppwm_matcher/models/user.rb]: right, but if it creates it without an email, it will hit the db again to add the email.

Comment (rwjblue) [inline lib/ppwm_matcher/models/user.rb]: OK, I get what you are saying.  `first_or_create` merges the two hashes and uses that to call `create`.

For a few examples and more detail see [here](http://api.rubyonrails.org/classes/ActiveRecord/Relation.html#method-i-first_or_create).

Comment (house9) [inline lib/ppwm_matcher/models/user.rb]: looks like `first_or_initialize` is the ticket - will push a commit shortly

```
    def self.find_or_create(email, github_user)
      user = where(:github_login => github_user.login).first_or_initialize(
        :gravatar_id   => github_user.gravatar_id,
        :github_login  => github_user.login
      )
      user.email = email
      user.save
      user
    end
```

File: lib/ppwm_matcher/models/user.rb (hunk 2 above)
Comment (house9) [inline lib/ppwm_matcher/models/user.rb]: `self.update_or_create` seems good?

**Pull request merged**